ambry-cloud/src/main/java/com/github/ambry/cloud/CloudStorageCompactor.java

```diff
@@ -57,141 +51,130 @@ public CloudStorageCompactor(CloudDestination cloudDestination, CloudConfig clou
-          break;
-        }
-      }
-      while (compactionInProgress > 0) {
-        try {
-          totalBlobsPurged += executorCompletionService.take().get();
-          compactedPartitionCount++;
-        } catch (ExecutionException ex) {
-          vcrMetrics.compactionFailureCount.inc();
-        }
-        compactionInProgress--;
+    HashMap<String, Future<Integer>> compactionTasks = new HashMap<>();
+      /*
+        Just submit the jobs and let the executor service handle the assignment and scheduling.
+        Note that partitions can be updated between compaction cycles when a new replica is added or an old one removed.
+        However, I do not know the behavior when a Collection is modified while it is being iterated.
+       */
+    for (PartitionId partitionId: partitions) {
+      try {
+        String partitionIdStr = partitionId.toPathString();
+        Future<Integer> future = executorService.submit(() -> cloudDestination.compactPartition(partitionIdStr));
+        compactionTasks.put(partitionIdStr, future);
+      } catch (Throwable throwable) {
+        vcrMetrics.compactionFailureCount.inc();
+        logger.error("[COMPACT] Failed to submit compaction task for partition-{} due to {}",
+            partitionId.toPathString(), throwable.getMessage());
       }
-      doneLatch.get().countDown();
-    } catch (Throwable th) {
-      logger.error("Hit exception running compaction task", th);
-    } finally {
-      long compactionTime = (System.currentTimeMillis() - compactionStartTime) / 1000;
-      logger.info("Purged {} blobs in {} partitions taking {} seconds", totalBlobsPurged, compactedPartitionCount,
-          compactionTime);
-    }
-    return totalBlobsPurged;
-  }
-
-  /**
-   * Purge the inactive blobs in the specified partitions.
-   * @param partition the {@link PartitionId} to compact.
-   * @return the total number of blobs purged in the partition.
-   */
-  private int compactPartition(PartitionId partition) throws CloudStorageException {
-    if (isShutDown()) {
-      logger.info("Skipping compaction due to shut down.");
-      return 0;
     }
 
-    String partitionPath = partition.toPathString();
-    logger.info("Beginning dead blob compaction for partition {}", partitionPath);
-
-    if (!partitions.contains(partition)) {
-      // Looks like partition was reassigned since the loop started, so skip it
-      logger.warn("Skipping compaction of Partition {} as the partition was reassigned", partition);
-      return 0;
-    }
-
-    try {
-      return cloudDestination.compactPartition(partitionPath);
-    } catch (CloudStorageException ex) {
-      logger.error("Compaction failed for partition {}", partitionPath, ex);
-      throw ex;
+    // Wait for completion or shutdown, don't use any timeouts
+    for (String partitionIdStr : compactionTasks.keySet()) {
+      // Although this is a finite loop, the future.get() is a blocking one, and we can't predict how long that'll be.
```

Comment:
The problem is that although this thread quit, the ExecutorService is still running and try to finish all the submitted 100 tasks.

Reply:
added an early shutdown

Reply:
Completely removed the "early shutdown" logic. Its just messy as hell.